File: backend/app/analysis.py
"""Infer athlete thresholds and weekly availability from recent Strava history.

These are deliberately simple, transparent heuristics meant to *seed* the
athlete profile from summary activity data (we don't pull per-second streams).
Every value is surfaced in the UI for the athlete to confirm or override, and
all of them feed TSS/zone math downstream.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass
from datetime import date, datetime, timedelta


@dataclass
class InferredProfile:
    ftp: float | None = None
    threshold_hr: int | None = None
    max_hr: int | None = None
    threshold_pace_run: float | None = None  # sec/km
    css_swim: float | None = None  # sec/100m
    weekly_hours_target: float | None = None
    basis: dict | None = None  # human-readable explanation of each value

    def as_dict(self) -> dict:
        return asdict(self)


def _parse_day(s: str) -> date:
    return datetime.fromisoformat(s.replace("Z", "+00:00")).date()


def infer_profile(
    activities: list[dict],
    *,
    today: date,
    threshold_window_days: int = 84,
    availability_window_days: int = 56,
) -> InferredProfile:
    """activities: dicts with keys sport, start_date, moving_time, distance,
    weighted_power, avg_power, avg_hr, max_hr."""
    th_cutoff = today - timedelta(days=threshold_window_days)
    av_cutoff = today - timedelta(days=availability_window_days)

    best_np = 0.0  # bike normalized power, sustained efforts
    fastest_run_pace: float | None = None  # sec/km (smaller is faster)
    fastest_swim_pace: float | None = None  # sec/100m
    best_threshold_hr = 0
    observed_max_hr = 0
    availability_seconds = 0.0
    basis: dict = {}

    for a in activities:
        try:
            day = _parse_day(a["start_date"])
        except (KeyError, ValueError):
            continue
        sport = a.get("sport")
        moving = a.get("moving_time") or 0
        distance = a.get("distance") or 0
        avg_hr = a.get("avg_hr")
        max_hr = a.get("max_hr")

        if max_hr:
            observed_max_hr = max(observed_max_hr, int(max_hr))

        # Availability: total moving time across all sports in recent window.
        if day >= av_cutoff:
            availability_seconds += moving

        if day < th_cutoff:
            continue

        sustained = moving >= 1200  # >= 20 minutes

        if sport == "Bike" and sustained:
            np = a.get("weighted_power") or a.get("avg_power")
            if np:
                best_np = max(best_np, float(np))
            if avg_hr:
                best_threshold_hr = max(best_threshold_hr, int(avg_hr))

        if sport == "Run" and sustained and distance > 0:
            pace = moving / (distance / 1000.0)  # sec/km
            if fastest_run_pace is None or pace < fastest_run_pace:
                fastest_run_pace = pace
            if avg_hr:
                best_threshold_hr = max(best_threshold_hr, int(avg_hr))

        if sport == "Swim" and distance >= 400:
            pace = moving / (distance / 100.0)  # sec/100m
            if fastest_swim_pace is None or pace < fastest_swim_pace:
                fastest_swim_pace = pace

    profile = InferredProfile()

    if best_np > 0:
        profile.ftp = round(best_np * 0.95, 0)  # ~95% of best sustained NP
        basis["ftp"] = f"95% of best sustained bike NP ({best_np:.0f}W) in last 12 weeks"

    if fastest_run_pace:
        # True threshold pace is slightly slower than a single best hard effort.
        profile.threshold_pace_run = round(fastest_run_pace * 1.04, 0)
        basis["threshold_pace_run"] = (
            f"fastest sustained run pace ({_fmt_pace_km(fastest_run_pace)}) +4%"
        )

    if fastest_swim_pace:
        profile.css_swim = round(fastest_swim_pace, 0)
        basis["css_swim"] = f"fastest sustained swim pace ({_fmt_pace_100(fastest_swim_pace)})"

    if best_threshold_hr:
        profile.threshold_hr = best_threshold_hr
        basis["threshold_hr"] = (
            f"max avg HR on sustained efforts ({best_threshold_hr} bpm)"
        )
    elif observed_max_hr:
        profile.threshold_hr = round(observed_max_hr * 0.92)
        basis["threshold_hr"] = f"~92% of observed max HR ({observed_max_hr} bpm)"

    if observed_max_hr:
        profile.max_hr = observed_max_hr

    weeks = availability_window_days / 7.0
    hours = availability_seconds / 3600.0
    if hours > 0:
        profile.weekly_hours_target = round(hours / weeks, 1)
        basis["weekly_hours_target"] = (
            f"avg of {hours:.1f}h over last {int(weeks)} weeks"
        )

    profile.basis = basis
    return profile


def _fmt_pace_km(sec_per_km: float) -> str:
    m, s = divmod(int(sec_per_km), 60)
    return f"{m}:{s:02d}/km"


def _fmt_pace_100(sec_per_100: float) -> str:
    m, s = divmod(int(sec_per_100), 60)
    return f"{m}:{s:02d}/100m"

File: backend-v2/src/main/java/io/gamov/irontrainer/athlete/Analysis.java
package io.gamov.irontrainer.athlete;

import io.gamov.irontrainer.activity.Activity;
import io.gamov.irontrainer.metrics.MetricsWrite;
import io.gamov.irontrainer.util.Iso;
import io.gamov.irontrainer.util.Py;
import io.gamov.irontrainer.util.PyJson;

import java.time.LocalDate;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/** Infer athlete thresholds from recent Strava summary history — faithful port of
 * app/analysis.py (infer_profile) + services.seed_profile_if_empty. Deliberately
 * simple heuristics that SEED an empty profile; every value is surfaced for the
 * athlete to confirm. JSON must match FastAPI: round(x,0)/round(x,1) → Py.round
 * (float), round()/int() → integer, `:.0f`/`:.1f` → Py.f0/Py.f1. */
public final class Analysis {

    private Analysis() {}

    static final int THRESHOLD_WINDOW_DAYS = 84;
    static final int AVAILABILITY_WINDOW_DAYS = 56;

    /** infer_profile with the default 84d threshold / 56d availability windows. */
    public static Map<String, Object> inferProfile(List<Activity> activities, LocalDate today) {
        return inferProfile(activities, today, THRESHOLD_WINDOW_DAYS, AVAILABILITY_WINDOW_DAYS);
    }

    public static Map<String, Object> inferProfile(List<Activity> activities, LocalDate today,
            int thresholdWindowDays, int availabilityWindowDays) {
        LocalDate thCutoff = today.minusDays(thresholdWindowDays);
        LocalDate avCutoff = today.minusDays(availabilityWindowDays);

        double bestNp = 0.0;                 // bike normalized power, sustained efforts
        Double fastestRunPace = null;        // sec/km (smaller is faster)
        Double fastestSwimPace = null;       // sec/100m
        int bestThresholdHr = 0;
        int observedMaxHr = 0;
        double availabilitySeconds = 0.0;

        for (Activity a : activities) {
            LocalDate day = Iso.parseDate(a.startDate);   // fromisoformat(...).date(), null on bad
            if (day == null) continue;       // Python: KeyError/ValueError → skip
            String sport = a.sport;
            double moving = a.movingTime == null ? 0 : a.movingTime;   // moving_time or 0
            double distance = a.distance == null ? 0 : a.distance;     // distance or 0
            Double avgHr = a.avgHr;
            Double maxHr = a.maxHr;

            if (maxHr != null && maxHr != 0.0) {                       // if max_hr:
                observedMaxHr = Math.max(observedMaxHr, (int) maxHr.doubleValue());
            }
            if (!day.isBefore(avCutoff)) availabilitySeconds += moving;  // day >= av_cutoff
            if (day.isBefore(thCutoff)) continue;                        // day < th_cutoff

            boolean sustained = moving >= 1200;                          // >= 20 min

            if ("Bike".equals(sport) && sustained) {
                Double np = (a.weightedPower != null && a.weightedPower != 0.0)
                        ? a.weightedPower : a.avgPower;                  // weighted_power or avg_power
                if (np != null && np != 0.0) bestNp = Math.max(bestNp, np);
                if (avgHr != null && avgHr != 0.0) {
                    bestThresholdHr = Math.max(bestThresholdHr, (int) avgHr.doubleValue());
                }
            }
            if ("Run".equals(sport) && sustained && distance > 0) {
                double pace = moving / (distance / 1000.0);             // sec/km
                if (fastestRunPace == null || pace < fastestRunPace) fastestRunPace = pace;
                if (avgHr != null && avgHr != 0.0) {
                    bestThresholdHr = Math.max(bestThresholdHr, (int) avgHr.doubleValue());
                }
            }
            if ("Swim".equals(sport) && distance >= 400) {
                double pace = moving / (distance / 100.0);              // sec/100m
                if (fastestSwimPace == null || pace < fastestSwimPace) fastestSwimPace = pace;
            }
        }

        Map<String, Object> basis = new LinkedHashMap<>();
        Map<String, Object> out = new LinkedHashMap<>();
        // InferredProfile.as_dict() field order: ftp, threshold_hr, max_hr,
        // threshold_pace_run, css_swim, weekly_hours_target, basis.
        Object ftp = null, thresholdHr = null, maxHr = null,
                thresholdPaceRun = null, cssSwim = null, weeklyHoursTarget = null;

        if (bestNp > 0) {
            ftp = Py.round(bestNp * 0.95, 0);                           // ~95% of best sustained NP
            basis.put("ftp", "95% of best sustained bike NP (" + Py.f0(bestNp)
                    + "W) in last 12 weeks");
        }
        if (fastestRunPace != null && fastestRunPace != 0.0) {
            thresholdPaceRun = Py.round(fastestRunPace * 1.04, 0);      // true threshold ~4% slower
            basis.put("threshold_pace_run", "fastest sustained run pace ("
                    + fmtPaceKm(fastestRunPace) + ") +4%");
        }
        if (fastestSwimPace != null && fastestSwimPace != 0.0) {
            cssSwim = Py.round(fastestSwimPace, 0);
            basis.put("css_swim", "fastest sustained swim pace (" + fmtPace100(fastestSwimPace) + ")");
        }
        if (bestThresholdHr > 0) {
            thresholdHr = (long) bestThresholdHr;
            basis.put("threshold_hr", "max avg HR on sustained efforts (" + bestThresholdHr + " bpm)");
        } else if (observedMaxHr > 0) {
            thresholdHr = Py.roundInt(observedMaxHr * 0.92);
            basis.put("threshold_hr", "~92% of observed max HR (" + observedMaxHr + " bpm)");
        }
        if (observedMaxHr > 0) {
            maxHr = (long) observedMaxHr;
        }
        double weeks = availabilityWindowDays / 7.0;
        double hours = availabilitySeconds / 3600.0;
        if (hours > 0) {
            weeklyHoursTarget = Py.round(hours / weeks, 1);
            basis.put("weekly_hours_target", "avg of " + Py.f1(hours)
                    + "h over last " + (int) weeks + " weeks");
        }

        out.put("ftp", ftp);
        out.put("threshold_hr", thresholdHr);
        out.put("max_hr", maxHr);
        out.put("threshold_pace_run", thresholdPaceRun);
        out.put("css_swim", cssSwim);
        out.put("weekly_hours_target", weeklyHoursTarget);
        out.put("basis", basis);
        return out;
    }

    /** seed_profile_if_empty: infer + persist thresholds ONLY when the athlete
     * has none yet (a first-connect convenience; no-op once any of ftp/threshold_hr/
     * threshold_pace_run/css_swim is set — NOT max_hr/weekly, matching Python).
     * Re-costs activities with the inferred thresholds (recompute_tss) but does
     * NOT rebuild the PMC — the caller (Strava sync / archive import) rebuilds
     * afterward. Returns the inferred map, or null when already seeded. Must run
     * inside the caller's transaction. */
    public static Map<String, Object> seedProfileIfEmpty(int aid, LocalDate today) {
        Athlete a = Athlete.findById(aid);
        if (a == null) return null;
        boolean already = Py.truthy(a.ftp) || Py.truthy(a.thresholdHr)
                || Py.truthy(a.thresholdPaceRun) || Py.truthy(a.cssSwim);
        if (already) return null;
        List<Activity> acts = Activity.list(
                "athleteId = ?1 and (isDuplicate = 0 or isDuplicate is null) order by startDate", aid);
        Map<String, Object> inferred = inferProfile(acts, today);
        saveInferred(a, inferred);          // fill blanks only
        MetricsWrite.recomputeTss(aid);     // re-cost only; caller rebuilds the PMC
        return inferred;
    }

    /** save_profile for inferred thresholds: sets only the non-null values
     * (never clears a field inference couldn't compute) and bumps updated_at.
     * The entity is managed inside the caller's transaction, so field mutation
     * flushes on commit — no explicit persist() needed. */
    public static void saveInferred(Athlete a, Map<String, Object> inferred) {
        if (a == null) return;
        boolean changed = false;
        Object v;
        if ((v = inferred.get("ftp")) != null) { a.ftp = ((Number) v).doubleValue(); changed = true; }
        if ((v = inferred.get("threshold_hr")) != null) { a.thresholdHr = ((Number) v).intValue(); changed = true; }
        if ((v = inferred.get("max_hr")) != null) { a.maxHr = ((Number) v).intValue(); changed = true; }
        if ((v = inferred.get("threshold_pace_run")) != null) { a.thresholdPaceRun = ((Number) v).doubleValue(); changed = true; }
        if ((v = inferred.get("css_swim")) != null) { a.cssSwim = ((Number) v).doubleValue(); changed = true; }
        if ((v = inferred.get("weekly_hours_target")) != null) { a.weeklyHoursTarget = ((Number) v).doubleValue(); changed = true; }
        if (changed) {
            a.updatedAt = PyJson.utcNowIso();   // _now_iso(): shared UTC ISO, matches every writer
        }
    }

    private static String fmtPaceKm(double secPerKm) {
        int v = (int) secPerKm;                      // int() truncates toward zero
        return String.format(Locale.ROOT, "%d:%02d/km", v / 60, v % 60);
    }

    private static String fmtPace100(double secPer100) {
        int v = (int) secPer100;
        return String.format(Locale.ROOT, "%d:%02d/100m", v / 60, v % 60);
    }
}
